examples/UNet/massachussets_road_segm.py

```diff
@@ -0,0 +1,340 @@
+
+    # run as consumer (read items from queue, in current thread)
+    item = queue.get()
+    while item is not sentinel:
+        yield item
+        queue.task_done()
+        item = queue.get()
+
+def prepare_dataset():
+    prep_folders()
+    prep_urls()
+    download_dataset()
+    # the dataset is now downloaded in the background. Once every few seconds we check if the download is done. We do
+    # this by checking whether tha last training image exists
+    while not os.path.isfile("data/training/map/99238675_15.tiff") and not os.path.isfile("data/training/sat_img/99238675_15.tiff"):
+        print "download seems to be running..."
+        sleep(5)
+    print "download done..."
+    try:
+        data_train, target_train = load_data("data/training")
+        data_valid, target_valid = load_data("data/validation")
+        data_test, target_test = load_data("data/test")
+        # loading np arrays is much faster than loading the images one by one every time
+        np.save("train_data.npy", data_train)
+        np.save("train_target.npy", target_train)
+        np.save("valid_data.npy", data_valid)
+        np.save("valid_target.npy", target_valid)
+        np.save("test_data.npy", data_test)
+        np.save("test_target.npy", target_test)
+    except:
+        print "something went wrong, maybe the download?"
+
+
+def plot_some_results(pred_fn, test_generator, BATCH_SIZE, PATCH_SIZE = 192):
+    fig_ctr = 0
+    ctr = 0
+    for data, seg in test_generator:
+        pred = pred_fn(data)
+        res = pred.argmax(-1).reshape(BATCH_SIZE, 1, PATCH_SIZE, PATCH_SIZE)
+        prob = pred.transpose(1, 0).reshape(2, BATCH_SIZE, PATCH_SIZE, PATCH_SIZE)[1]
+        for d, s, r, p in zip(data, seg, res, prob):
+            plt.figure(figsize=(12, 6))
+            plt.subplot(1, 4, 1)
+            plt.imshow(d.transpose(1,2,0))
+            plt.subplot(1, 4, 2)
+            plt.imshow(s[0])
+            plt.subplot(1, 4, 3)
+            plt.imshow(r[0])
+            plt.subplot(1, 4, 4)
+            plt.imshow(p)
+            plt.savefig("road_segmentation_result_%03.0f.png"%fig_ctr)
+            plt.close()
+            fig_ctr += 1
+        ctr += 1
+        if ctr > 5:
+            break
+
+
+def main():
+    # only download dataset once. This takes a while.
+    if not os.path.isfile("test_target.npy"):
+        # heuristic that I included to make sure the dataset is only donwloaded and prepared once
+        prepare_dataset()
+
+    # set some hyper parameters. You should not have to touch anything if you have 4GB or more VRAM
+    BATCH_SIZE = 24
+    N_EPOCHS = 30
+    N_BATCHES_PER_EPOCH = 100
+    N_BATCHES_PER_EPOCH_valid = 15
+    PATCH_SIZE = 128+64
+
+    # load the prepared data. They have been converted to np arrays because they are much faster to load than single image files.
+    # you will need some ram in order to have everything in memory.
+    # TODO is it worth the effort to refactor everything so that we use memmaps?
```

Comment:
No need to refactor anything: Just add `mmap_mode='r'` to each `np.load` call. Probably this should be optional, though. If you have enough RAM, loading them before will be faster because then the data is read sequentially, not in random order as requested during training.